### PR TITLE
server: improve the addMetrics method performance

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

### What problem does this PR solve? 

1. The `addMetrics` method is a hot path in TiDB and will be executed in every query. It uses `switch .. case` to find the `Counter` and add metrics to monitor system. It's not an efficient way.
2. Every metric use two global variables  (ok/error), which is global variable abuse.

See: https://godbolt.org/z/czr__4

### What is changed and how it works?

This PR makes the following change.

1. Group related metrics into two global variables by status (ok/error) and save them in an array. It's fast to get `Counter` by `cmd` (index).
2. Use a hash map to save statement type related metrics. And find the `Observer` from map is fast than `switch .. case` in common case.

### Check List 

Tests 

 - Unit test
 - Integration test

Release note

 - No need to add to release note.

### CODE REVIEW

**Comment by ngaut** [comment]
Do you have bench results?

File: server/conn.go [hunk 1]
```diff
@@ -713,132 +718,39 @@ func errStrForLog(err error) string {
...
-	case "Select":
-		queryDurationHistogramSelect.Observe(time.Since(startTime).Seconds())
-	case "Execute":
-		queryDurationHistogramExecute.Observe(time.Since(startTime).Seconds())
-	case "Set":
-		queryDurationHistogramSet.Observe(time.Since(startTime).Seconds())
-	case metrics.LblGeneral:
-		queryDurationHistogramGeneral.Observe(time.Since(startTime).Seconds())
-	default:
+	observer, found := queryDurationHistogram[sqlType]
```
Comment (lysu) [inline server/conn.go]: how about refactor `sqlType` to be int first, then here also can use array :D

**Comment by lonng** [comment]
@ngaut 
I did a simple benchmark and got two conclusions.
1. If you use hashmap, performance regression will occur

    ```
    BenchmarkSwitchCase-8         	287540623	         4.18 ns/op
    BenchmarkArrayAndHashMap-8    	70893422	        16.3 ns/op
    BenchmarkSwitchCase2-8        	330713746	         3.60 ns/op
    BenchmarkArrayAndHashMap2-8   	68912976	        16.1 ns/op
    BenchmarkSwitchCase3-8        	333756572	         3.56 ns/op
    BenchmarkArrayAndHashMap3-8   	70810513	        16.2 ns/op
    ```
2. If you only use an array, there will be 2-4x performance improvement

    ```
    BenchmarkSwitchCase-8         	161118134	         7.48 ns/op
    BenchmarkArrayAndHashMap-8    	566121494	         2.35 ns/op
    BenchmarkSwitchCase2-8        	198848967	         5.75 ns/op
    BenchmarkArrayAndHashMap2-8   	400275949	         2.64 ns/op
    BenchmarkSwitchCase3-8        	451752652	         2.51 ns/op
    BenchmarkArrayAndHashMap3-8   	666522830	         1.79 ns/op
    ```

See snippet: https://gist.github.com/lonng/339c17800a20fe48998d552d9ee1d535

So I have reverted the hashmap part.

**Comment by lonng** [comment]
@lysu @jackysp PTAL

**Comment by cfzjywxk** [comment]
LGTM

**Pull request merged**